### PR TITLE
Add nix shell support

### LINKED ISSUE #29: NIx support
I could be useful to have a `shell.nix` file that sets up an environment with `haskellings` already in the path and maybe setup editor integration.

### PR DESCRIPTION
**Pull request by jakeisnt**
closes #29

### CODE REVIEW

**Comment by jakeisnt** [comment]
Looks like this is working! To test it out, follow the instructions in the README.

Note that the build that Nix uses does not run tests (`doCheck` in `haskellings.nix` is set to false) but as dependencies are pinned, you know that you're getting the same package versions and configurations every time - so as long as `haskellings` works at runtime this is not an issue.

**Comment by jakeisnt** [comment]
I don't have substantial knowledge of `haskellings` but I'm not sure if the program is working properly when compiled this way. I'm able to run the exercises, but it seems like the output of `haskellings` on specific exercises is being cached for some reason. Make sure to test this rigorously before merging.

File: .gitignore [hunk 1]
```diff
@@ -5,3 +5,4 @@ generated_files/
 .config/
 tests/test_gen/
 config.yaml
+result
```
Comment (RosarioPulella) [inline .gitignore:8]:
```suggestion
result
.direnv/
```

If we are also adding a `.envrc` we should also have `.direnv/`(direnv's cashe) in `.gitignore`

Comment (jakeisnt) [reply .gitignore:8]: `use nix` doesn't use direnv's cache, afaik, but `use flake` does. wouldn't be bad to add it either way.

File: flake.lock [hunk 2]
```diff
@@ -0,0 +1,42 @@
+{
```
Comment (MondayMorningHaskell) [inline flake.lock:1]: Are these Flake and Env files (`.envrc`, `flake.lock` and `flake.nix`) required for this PR? I just noticed the README mention that Nix Flakes is optional.

I suppose I have a small aesthetic interest in keeping the project root less cluttered but if it's required then it's fine.

**Review by MondayMorningHaskell — COMMENTED** [review]
I'm not able to test this at the moment, running into an install issue with WSL Ubuntu. Don't suppose you've heard of the error `multi-user installation requires systemd on Linux`? Might try posting an issue on Nix repo if not.

In a couple weeks I'll be able to give this a try on my Mac but currently traveling without it 😬

**Comment by jakeisnt** [comment]
Unfortunately, I don't have any experience using Nix  on WSL. It's possible you can just install `systemd` via `apt`, but beyond that I'm not sure what to try.

Those files are not strictly required, but here's what they're for:

## Direnv

If you have the program direnv installed, it will automatically run what's in the .envrc file. This means that no commands have to be typed in to enter the nix shell environment; without it, you have to type `nix-shell` every time you enter the repository. 

As most editors can't just run this command, adding a direnv plugin to the editor allows the editor to automatically load all of the Nix dependencies from the Nix shell into the local environment when the project is opened without having to run nix-shell then start the editor inside of that local environment. 

If someone didn't have a language server installed globally, for example, they would either need to enter the nix shell then start the command for their editor or use direnv to make that language server available to the editor. 

Direnv support can be removed without impacting Nix, but for those reasons I believe it's important to provide.

## Nix Flakes

The flake.nix and flake.lock files enable support for an experimental Nix feature, nix flakes, that's stable and is on track to be made the canonical way to use Nix soon. They provide a structured way to precisely pin dependencies and cache builds to ensure that you get exactly the same shell every time. (You'll notice that one of the other nix files just calls a `fetchTarball` function with a URL; this is the easiest way to pin the exact versions of the nixpkgs repo, and therefore the packages installed, without flakes).

Aside from potentially futureproofing the repository, the flake version of `nix-shell` (`nix-develop`) is much faster and will completely cache the build. `nix-shell` here will recompile `haskellings` every time you enter the repository, but `nix develop` will not. It's what I prefer to use for my projects. 

As set up here, the shell environments from `nix develop` and `nix-shell` should be identical. Without `flake.nix` and `flake.lock`, `nix-shell` will still work just fine, but you won't get the instant load times or build caching that flakes can enable. 

I chose to make this optional because some users may just not want to enable flakes - it's more work than running the one line install command on the Nix website and some may be averse to its current status as an experimental feature.


All of that said, I'd be cool with removing any or all of the files you mentioned from the PR if you aren't interested in including them yet!

File: README.md [hunk 3]
```diff
@@ -27,6 +27,17 @@ You can also run any `haskellings` command from the project directory using `sta
 >> stack exec -- haskellings run Types1
 ```
 
+### Using Nix
```
Comment (MondayMorningHaskell) [inline README.md:30]: Let's add in this section that the usage is still a bit experimental.

**Review by MondayMorningHaskell — COMMENTED** [review]
Yea I'd say let's exclude the Flake and `.envrc` files (or add them to `.gitignore`) before merging.

File: README.md (hunk 3 above)
Comment (jakeisnt) [reply README.md]: Just revised the passage! Let me know how it looks now.

**Comment by RosarioPulella** [comment]
Still did not get a chance to test this out yet I will see if i can soon. I do like that its more minimal now.

I think there might be room for improvement in a future PR, like maybe using stack's nix integration or nixpkgs' `buildStackProject` function.

Some useful docks 
https://haskell4nix.readthedocs.io/nixpkgs-users-guide.html#how-to-build-a-haskell-project-using-stack

File: README.md (hunk 3 above)
Comment (jhb563) [reply README.md]: Looks good!

Comment (MondayMorningHaskell) [reply README.md]: Looks good!

**Comment by MondayMorningHaskell** [comment]
I'm happy with the state of the code. Hopefully I can get a chance to test it out soon! But if @RosarioPulella can test and verify that works I think that's good enough to merge on a provisional basis.

**Comment by RosarioPulella** [comment]
Yeah I did some testing and it seams like its working, it builds haskellings with the right ghc, and it runs just find, but I am not totally sure its using the right stackage snapshot. If need be I can improve it further later on after some research, but this seems like a good start.

**Review by RosarioPulella — APPROVED** [review]

**Comment by RosarioPulella** [comment]
@MondayMorningHaskell I would say this is ready if you want to merge it.

**Pull request merged**